index.d.ts

```diff
@@ -0,0 +1,204 @@
+  result: string; // TODO: Should we also document number?
+
+  startSpan: StartSpanFn;
+  ensureParentId (): string;
+  end: EndTransactionFn;
+}
+
+declare class Span extends GenericSpan {
+  // TODO: The following should not be documented right? constructor(), customStackTrace(), setDbContext()
+  transaction: Transaction;
+  name: string | null; // TODO: Should we allow null?
+
+  end (endTime?: number): void;
+}
+
+interface AgentConfigOptions {
+  abortedErrorThreshold?: string | number; // TODO: Do we officially want to support numbers?
+  active?: boolean;
+  apiRequestSize?: string | number; // TODO: Do we officially want to support numbers?
+  apiRequestTime?: string | number; // TODO: Do we officially want to support numbers?
+  asyncHooks?: boolean;
+  captureBody?: CaptureBody;
+  captureErrorLogStackTraces?: CaptureErrorLogStackTraces;
+  captureExceptions?: boolean;
+  captureHeaders?: boolean;
+  captureSpanStackTraces?: boolean;
+  containerId?: string;
+  disableInstrumentations?: string | string[]; // TODO: Do we officially want to support strings?
+  errorMessageMaxLength?: string | number; // TODO: Do we officially want to support numbers?
+  errorOnAbortedRequests?: boolean;
+  filterHttpHeaders?: boolean;
+  frameworkName?: string;
+  frameworkVersion?: string;
+  hostname?: string;
+  ignoreUrls?: Array<string | RegExp>;
+  ignoreUserAgents?: Array<string | RegExp>;
+  instrument?: boolean;
+  kubernetesNamespace?: string;
+  kubernetesNodeName?: string;
+  kubernetesPodName?: string;
+  kubernetesPodUID?: string;
+  logLevel?: LogLevel;
+  logger?: Logger;
+  metricsInterval?: string | number; // TODO: Do we officially want to support numbers?
+  payloadLogFile?: string; // TODO: Should we advertise this API?
+  secretToken?: string;
+  serverTimeout?: string | number; // TODO: Do we officially want to support numbers?
+  serverUrl?: string;
+  serviceName?: string;
+  serviceVersion?: string;
+  sourceLinesErrorAppFrames?: number;
+  sourceLinesErrorLibraryFrames?: number;
+  sourceLinesSpanAppFrames?: number;
+  sourceLinesSpanLibraryFrames?: number;
+  stackTraceLimit?: number;
+  transactionMaxSpans?: number;
+  transactionSampleRate?: number;
+  verifyServerCert?: boolean;
+}
+
+interface CaptureErrorOptions {
+  request?: IncomingMessage; // TODO: Currently not documented - should we expose this?
+  response?: ServerResponse; // TODO: Currently not documented - should we expose this?
+  timestamp?: number;
+  handled?: boolean; // TODO: Currently not documented - should we expose this?
+  user?: UserObject;  
+  tags?: Tags; // TODO: Currently not documented
+  custom?: object;
+  message?: string;
+}
+
+interface Tags {
+  [key: string]: TagValue;
+}
+
+interface UserObject {
+  id?: string | number | null | undefined;
+  username?: string | null | undefined;
+  email?: string | null | undefined;
+}
+
+interface ParameterizedMessageObject {
+  message: string;
+  params: Array<any>;
+}
+
+interface Logger {
```

Comment:
Pay attention that often a logger has also other methods, such as `child` for example.

Reply:
Ah ok, is there a way to say that it has these methods + allow any other method?

Reply:
I would try with something like following.
```ts
type anyFn = (...args: any[]) => any;
export interface Logger {
  fatal (msg: string, ...args: any[]): void;
  fatal (obj: {}, msg?: string, ...args: any[]): void;
  ...
  [key: string]: anyFn;
}
```
But I always had some issue with `[key: string]` usage. Mainly because I don't really understand what ts expects from me.

Reply:
Ah that's simple. Thanks. It's just saying that you can have any property in the `Logger` interface of the type defined - in this case `anyFn`.

I also just found this in the TypeScript docs:

![image](https://user-images.githubusercontent.com/10602/54493288-649d3d00-48ce-11e9-9550-ed6084d58eb9.png)